src/common/framework/tree.ts

```diff
@@ -1,82 +1,334 @@
-  const parts = path.split('/');
-  if (parts.length > 1) {
-    let partial = parts[0] + '/';
-    yield partial;
-    for (let i = 1; i < parts.length - 1; ++i) {
-      partial += parts[i] + '/';
-      yield partial;
+export interface TestTreeLeaf {
+  readonly query: TestQuerySingleCase;
+  readonly run: RunFn;
+}
+
+export type TestTreeNode = TestSubtree | TestTreeLeaf;
+
+export class TestTree {
+  readonly root: TestSubtree;
+
+  constructor(root: TestSubtree) {
+    this.root = root;
+  }
+
+  iterateCollapsedQueries(): IterableIterator<TestQuery> {
+    return TestTree.iterateSubtreeCollapsedQueries(this.root);
+  }
+
+  iterateLeaves(): IterableIterator<TestTreeLeaf> {
+    return TestTree.iterateSubtreeLeaves(this.root);
+  }
+
+  toString(): string {
+    return TestTree.subtreeToString('(root)', this.root, '');
+  }
+
+  static *iterateSubtreeCollapsedQueries(subtree: TestSubtree): IterableIterator<TestQuery> {
+    for (const [, child] of subtree.children) {
+      if ('children' in child && !child.collapsible) {
+        yield* TestTree.iterateSubtreeCollapsedQueries(child);
+      } else {
+        yield child.query;
+      }
     }
-    // Path ends in '/' (so is a README).
-    if (parts[parts.length - 1] === '') {
-      return;
+  }
+
+  static *iterateSubtreeLeaves(subtree: TestSubtree): IterableIterator<TestTreeLeaf> {
+    for (const [, child] of subtree.children) {
+      if ('children' in child) {
+        yield* TestTree.iterateSubtreeLeaves(child);
+      } else {
+        yield child;
+      }
     }
   }
-  yield path + terminator;
-}
 
-export function treeFromFilterResults(
-  log: Logger,
-  listing: IterableIterator<TestFilterResult>
-): FilterResultTreeNode {
-  function getOrInsert(n: FilterResultTreeNode, k: string): FilterResultTreeNode {
-    const children = n.children!;
-    if (children.has(k)) {
-      return children.get(k)!;
+  static subtreeToString(name: string, tree: TestTreeNode, indent: string): string {
+    const collapsible = 'run' in tree ? '>' : tree.collapsible ? '+' : '-';
+    let s =
+      indent +
+      `${collapsible} ${JSON.stringify(name)} => ` +
+      `${tree.query}        ${JSON.stringify(tree.query)}`;
+    if ('children' in tree) {
+      if (tree.description !== undefined) {
+        s += indent + `\n    | ${JSON.stringify(tree.description)}`;
+      }
+
+      for (const [name, child] of tree.children) {
+        s += '\n' + TestTree.subtreeToString(name, child, indent + '  ');
+      }
     }
-    const v = { children: new Map() };
-    children.set(k, v);
-    return v;
+    return s;
   }
+}
 
-  const tree = { children: new Map() };
-  for (const f of listing) {
-    const files = getOrInsert(tree, f.id.suite + ':');
-    if (f.id.path === '') {
-      // This is a suite README.
-      files.description = f.spec.description.trim();
+// TODO: Consider having subqueriesToExpand actually impact the depth-order of params in the tree.
+export async function loadTreeForQuery(
```

Comment:
This could use a comment describing what `subqueriesToExpand` is used for and overall what the function does. IIUC, we load the tests in `queryToLoad` but we split out ones in `subqueriesToExpand` because those are used for expectations. Splitting them out allows running separately and minimizing the number of queries?